advanced_hw_2.py

```diff
@@ -0,0 +1,412 @@
+            yield (key, self[key])
+
+dct = MyDict({"a": 1, "b": 2, "c": 3, "d": 25})
+for key, value in dct:
+    print(key, value)
+
+for key, value in dct.items():
+    print(key, value)
+
+for key in dct.keys():
+    print(key)
+
+dct["c"] + dct["d"]
+
+"""# Задание 2 (2 балла)
+
+Напишите функцию `iter_append`, которая "добавляет" новый элемент в конец итератора, возвращая итератор, который включает изначальные элементы и новый элемент. Итерироваться по итератору внутри функции нельзя, то есть вот такая штука не принимается
+```python
+def iter_append(iterator, item):
+    lst = list(iterator) + [item]
+    return iter(lst)
+```
+
+**Модули использовать нельзя**
+"""
+
+def iter_append(iterator, item):
+    yield from iterator
+    yield item 
+    pass
+
+my_iterator = iter([1, 2, 3])
+new_iterator = iter_append(my_iterator, 4)
+
+for element in new_iterator:
+    print(element)
+
+"""# Задание 3 (5 баллов)
+
+Представим, что мы установили себе некотурую библиотеку, которая содержит в себе два класса `MyString` и `MySet`, которые являются наследниками `str` и `set`, но также несут и дополнительные методы.
+
+Проблема заключается в том, что библиотеку писали не очень аккуратные люди, поэтому получилось так, что некоторые методы возвращают не тот тип данных, который мы ожидаем. Например, `MyString().reverse()` возвращает объект класса `str`, хотя логичнее было бы ожидать объект класса `MyString`.
+
+Найдите и реализуйте удобный способ сделать так, чтобы подобные методы возвращали экземпляр текущего класса, а не родительского. При этом **код методов изменять нельзя**
+
+**+3 дополнительных балла** за реализацию того, чтобы **унаследованные от `str` и `set` методы** также возвращали объект интересующего нас класса (то есть `MyString.replace(..., ...)` должен возвращать `MyString`). **Переопределять методы нельзя**
+
+**Модули использовать нельзя**
+"""
+
+def return_other_type(method):
+    def wrapper(*args, **kwargs):
+        result = method(*args, **kwargs)
+        if isinstance(result, str):
+            return args[0].__class__(result)
+        elif isinstance(result, set):
+            return args[0].__class__(result)
+        return result
+    return wrapper
+
+class MyString(str):
+    @return_other_type
+    def reverse(self):
+        return self[::-1]
+    
+    @return_other_type
+    def make_uppercase(self):
+        return "".join([chr(ord(char) - 32) if 97 <= ord(char) <= 122 else char for char in self])
+    
+    @return_other_type
+    def make_lowercase(self):
+        return "".join([chr(ord(char) + 32) if 65 <= ord(char) <= 90 else char for char in self])
+    
+    @return_other_type
+    def capitalize_words(self):
+        return " ".join([word.capitalize() for word in self.split()])
+    
+    
+class MySet(set):
+    @return_other_type
+    def is_empty(self):
+        return len(self) == 0
+    
+    @return_other_type
+    def has_duplicates(self):
+        return len(self) != len(set(self))
```

Comment:
Так как тут возвращается `bool`, декоратор не нужен
```suggestion
    def is_empty(self):
        return len(self) == 0
    
    def has_duplicates(self):
        return len(self) != len(set(self))
```